From <commit> Mon Sep 17 00:00:00 2001
From: <author>
Date: 2026-08-13
Subject: [PATCH] s390x: emit sqdbr for square root and lrvgr for byte swap

Both were going out to C. Sequences and operand semantics from the
Principles of Operation (SA22-7832), pages 19-40 and 7-328.
---
 asmcomp/s390x/CSE.ml       |  2 +-
 asmcomp/s390x/arch.ml      |  6 ++++++
 asmcomp/s390x/arch.mli     |  2 ++
 asmcomp/s390x/emit.mlp     | 20 +++++++++++++++++++-
 asmcomp/s390x/selection.ml | 31 ++++++++++++++++++++++++++++++-
 5 files changed, 58 insertions(+), 3 deletions(-)

diff --git a/asmcomp/s390x/CSE.ml b/asmcomp/s390x/CSE.ml
index dcefc4dc72..f1f0f9f7e4 100644
--- a/asmcomp/s390x/CSE.ml
+++ b/asmcomp/s390x/CSE.ml
@@ -28,7 +28,7 @@ inherit cse_generic as super
 
 method! class_of_operation op =
   match op with
-  | Ispecific(Imultaddf | Imultsubf) -> Op_pure
+  | Ispecific(Imultaddf | Imultsubf | Isqrtf | Ibswap _) -> Op_pure
   | _ -> super#class_of_operation op
 
 method! is_cheap_operation op =
diff --git a/asmcomp/s390x/arch.ml b/asmcomp/s390x/arch.ml
index e1d0a2cc03..e88b5dbc5c 100644
--- a/asmcomp/s390x/arch.ml
+++ b/asmcomp/s390x/arch.ml
@@ -33,6 +33,8 @@ let command_line_options =
 type specific_operation =
     Imultaddf                           (* multiply and add *)
   | Imultsubf                           (* multiply and subtract *)
+  | Isqrtf                              (* floating-point square root *)
+  | Ibswap of int                       (* endianness conversion *)
 
 (* Addressing modes *)
 
@@ -82,6 +84,10 @@ let print_specific_operation printreg op ppf arg =
   | Imultsubf ->
       fprintf ppf "%a *f %a -f %a"
         printreg arg.(0) printreg arg.(1) printreg arg.(2)
+  | Isqrtf ->
+      fprintf ppf "sqrtf %a" printreg arg.(0)
+  | Ibswap n ->
+      fprintf ppf "bswap_%i %a" n printreg arg.(0)
 
 (* Specific operations that are pure *)
 
diff --git a/asmcomp/s390x/arch.mli b/asmcomp/s390x/arch.mli
index 4b8f6dc2d4..defafedde2 100644
--- a/asmcomp/s390x/arch.mli
+++ b/asmcomp/s390x/arch.mli
@@ -27,6 +27,8 @@ val command_line_options : (string * Arg.spec * string) list
 type specific_operation =
     Imultaddf                           (* multiply and add *)
   | Imultsubf                           (* multiply and subtract *)
+  | Isqrtf                              (* floating-point square root *)
+  | Ibswap of int                       (* endianness conversion *)
 
 (* Addressing modes *)
 
diff --git a/asmcomp/s390x/emit.mlp b/asmcomp/s390x/emit.mlp
index 0e37e31371..d6edcc0243 100644
--- a/asmcomp/s390x/emit.mlp
+++ b/asmcomp/s390x/emit.mlp
@@ -284,6 +284,7 @@ let name_for_floatop2 = function
 let name_for_specific = function
     Imultaddf -> "madbr"
   | Imultsubf -> "msdbr"
+  | _ -> Misc.fatal_error "Emit.Ispecific"
 
 (* Output the assembly code for an instruction *)
 
@@ -589,7 +590,24 @@ let emit_instr env i =
         `	cgdbr	{emit_reg i.res.(0)}, 5, {emit_reg i.arg.(0)}\n`
     | Lop(Iopaque) ->
         assert (i.arg.(0).loc = i.res.(0).loc)
-    | Lop(Ispecific sop) ->
+    | Lop(Ispecific Isqrtf) ->
+        `	sqdbr	{emit_reg i.res.(0)}, {emit_reg i.arg.(0)}\n`
+    | Lop(Ispecific (Ibswap 16)) ->
+        (* Shifting the halfword up to the top means the full 64-bit reverse
+           brings it back down zero-extended, which is what the C version
+           returns. *)
+        `	sllg	{emit_reg i.res.(0)}, {emit_reg i.arg.(0)}, 48(%r0)\n`;
+        `	lrvgr	{emit_reg i.res.(0)}, {emit_reg i.res.(0)}\n`
+    | Lop(Ispecific (Ibswap 32)) ->
+        (* lrvr leaves bits 0-31 alone (PoP 7-328), and the C version returns
+           a sign-extended int32, so lgfr supplies the top half. *)
+        `	lrvr	{emit_reg i.res.(0)}, {emit_reg i.arg.(0)}\n`;
+        `	lgfr	{emit_reg i.res.(0)}, {emit_reg i.res.(0)}\n`
+    | Lop(Ispecific (Ibswap 64)) ->
+        `	lrvgr	{emit_reg i.res.(0)}, {emit_reg i.arg.(0)}\n`
+    | Lop(Ispecific (Ibswap _)) ->
+        Misc.fatal_error "Emit.Ibswap: unsupported width"
+    | Lop(Ispecific (Imultaddf | Imultsubf as sop)) ->
         assert (i.arg.(2).loc = i.res.(0).loc);
         let instr = name_for_specific sop in
         `	{emit_string instr}	{emit_reg i.res.(0)}, {emit_reg i.arg.(0)}, {emit_reg i.arg.(1)}\n`
diff --git a/asmcomp/s390x/selection.ml b/asmcomp/s390x/selection.ml
index cce96959be..c0cfea85be 100644
--- a/asmcomp/s390x/selection.ml
+++ b/asmcomp/s390x/selection.ml
@@ -52,7 +52,7 @@ let pseudoregs_for_operation op arg res =
   (* Two-address binary operations: arg.(0) and res.(0) must be the same *)
   | Iintop(Iadd|Isub|Imul|Iand|Ior|Ixor)  | Iaddf|Isubf|Imulf|Idivf ->
       ([|res.(0); arg.(1)|], res)
-  | Ispecific _ ->
+  | Ispecific (Imultaddf | Imultsubf) ->
     ( [| arg.(0); arg.(1); res.(0) |], [| res.(0) |])
   (* One-address unary operations: arg.(0) and res.(0) must be the same *)
   |  Iintop_imm((Imul|Iand|Ior|Ixor), _) -> (res, res)
@@ -62,6 +62,12 @@ let pseudoregs_for_operation op arg res =
 let is_immediate n = n <= 0x7FFF_FFFF && n >= -0x8000_0000
 let is_immediate_logical n = n <= 0xFFFF_FFFF && n >= 0
 
+(* If you update [inline_ops], you may need to update [is_simple_expr] and/or
+   [effects_of], below. *)
+let inline_ops =
+  [ "sqrt"; "caml_bswap16_direct"; "caml_int32_direct_bswap";
+    "caml_int64_direct_bswap"; "caml_nativeint_direct_bswap" ]
+
 class selector = object (self)
 
 inherit Selectgen.selector_generic as super
@@ -100,9 +106,32 @@ method! select_operation op args dbg =
       (Ispecific Imultaddf, [arg1; arg2; arg3])
   | (Csubf, [Cop(Cmulf, [arg1; arg2], _); arg3]) ->
       (Ispecific Imultsubf, [arg1; arg2; arg3])
+  (* Recognize square root *)
+  | (Cextcall("sqrt", _, _, _), [arg]) ->
+      (Ispecific Isqrtf, [arg])
+  (* Recognize byte swaps *)
+  | (Cextcall("caml_bswap16_direct", _, _, _), _) ->
+      (Ispecific (Ibswap 16), args)
+  | (Cextcall("caml_int32_direct_bswap", _, _, _), _) ->
+      (Ispecific (Ibswap 32), args)
+  | (Cextcall(("caml_int64_direct_bswap"|"caml_nativeint_direct_bswap"),
+              _, _, _), _) ->
+      (Ispecific (Ibswap 64), args)
   | _ ->
       super#select_operation op args dbg
 
+method! is_simple_expr = function
+  (* inlined floating-point ops are simple if their arguments are *)
+  | Cop(Cextcall (fn, _, _, _), args, _) when List.mem fn inline_ops ->
+      List.for_all self#is_simple_expr args
+  | e -> super#is_simple_expr e
+
+method! effects_of e =
+  match e with
+  | Cop(Cextcall (fn, _, _, _), args, _) when List.mem fn inline_ops ->
+      Selectgen.Effect_and_coeffect.join_list_map args self#effects_of
+  | e -> super#effects_of e
+
 method! insert_op_debug env op dbg rs rd =
   try
     let (rsrc, rdst) = pseudoregs_for_operation op rs rd in
-- 
2.51.0.windows.2

